ModelCatalogueCorePluginTestApp/src/groovy/org/modelcatalogue/core/mysql-procedures.sql

```diff
@@ -7,30 +7,48 @@ DETERMINISTIC
     BEGIN
         IF LENGTH(CurrentSet) = 0 OR CurrentSet is NULL THEN
             RETURN NextValues;
+        ELSEIF LENGTH(NextValues) = 0 OR NextValues is NULL THEN
+            RETURN CurrentSet;
+            # AddToSet should be symmetrical
         ELSE
             RETURN CONCAT(CurrentSet,',',NextValues);
         END IF;
     END $$
 
 DROP FUNCTION IF EXISTS GetAllDestinations $$
 
+/*
+ GetAllDestinations is basically a graph search in SQL! It finds all CatalogueElements z such that
+ there is x in InitialQueue where x =[DescendType]=>* y =[LeafType]=> z.
+ x =[rel]=>* y means you can reach y from x in 0 or more steps of relations rel.
+ The typical example is InitialQueue is the top level data classes of a model (accessed using, of course, GetTopLevelDataClasses),
+ DescendType is Hierarchy, which relates a class and its child class,
+ and LeafType is Containment (:hierarchytypeId and :containmentTypeId), which relates a data class and a data element
+ that it contains.
+ This is used in the DataElementService, DataModelService, and DataTypeService.
+ The DescendType may cross data models, which happens when a data model imports another data model.
+ */
 CREATE FUNCTION GetAllDestinations (InitialQueue MEDIUMTEXT, DescendType LONG, LeafType LONG) RETURNS MEDIUMTEXT
 DETERMINISTIC
     BEGIN
-
+        # Note that sets and queues such as InitialQueue and processed are represented as comma-separated lists.
         DECLARE rv,q,queue,queue_children,leaves, processed,data_models_with_containment MEDIUMTEXT;
         DECLARE queue_length,front_id,pos LONG;
 
         SET data_models_with_containment = (select group_concat(DISTINCT src.data_model_id) from relationship r
             join catalogue_element src on src.id = r.source_id
             join catalogue_element dest on dest.id = r.destination_id
-            where r.relationship_type_id = LeafType or r.relationship_type_id = DescendType and src.data_model_id <> dest.data_model_id);
-        SET rv = '';
+            where r.relationship_type_id = LeafType or
+                  (r.relationship_type_id = DescendType and src.data_model_id <> dest.data_model_id));
+
+        SET rv = ''; # This is the result set
         SET queue = InitialQueue;
         SET queue_length = LENGTH(InitialQueue) - LENGTH(REPLACE(InitialQueue, ',', '')) + 1;
+        # a clever trick for finding the length of a comma separated list
```

Comment:
I think to put the comments to the next line after the statement makes the association difficult. I would prefer format

```
# this is a comment
this is a line commented
```